ZenPacks/zenoss/OpenStackInfrastructure/datamaps.py

```diff
@@ -0,0 +1,201 @@
+##############################################################################
+#
+# Copyright (C) Zenoss, Inc. 2019, all rights reserved.
+#
+# This content is made available according to terms specified in
+# License.zenoss under the directory where your Zenoss product is installed.
+#
+##############################################################################
+
+from time import time
+
+from Products.DataCollector.plugins.DataMaps import ObjectMap
+
+
+class ConsolidatingObjectMapQueue(object):
+    """
+    Provide a black box which you can insert object maps into, and then
+    drain object maps out of.  The maps will be held for a period of time
+    before they are released, and incremental changes aggregated into a smaller
+    set of effective changes, when possible, according to the following rules:
+
+    * If a component is created by a map (_add=True), hold all updates
+      to this component for a period of time (shortlived_seconds).  If a
+      map comes in during that time to delete that same component, discard
+      all maps for it and don't waste zenhub's time.
+
+    * If a component is deleted (.remove=True), discard any other pending
+      object maps for this component, and release the remove datamap immediately.
+      In addition, store the component ID in a blacklist for
+      (delete_blacklist_seconds) seconds.
+
+    * If any objmap comes in for a blacklisted component, drop it.
+
+    * For all other updates, build an aggregate datamap, overlaying each
+      new datamap's changes onto it, until its age reaches
+      update_consolidate_seconds, then release it for consumption.
+    """
+
+    def __init__(self, shortlived_seconds=40, delete_blacklist_seconds=600, update_consolidate_seconds=40):
+        self.shortlived_seconds = shortlived_seconds
+        self.delete_blacklist_seconds = delete_blacklist_seconds
+        self.update_consolidate_seconds = update_consolidate_seconds
+        self.held_objmaps = {}
+        self.blacklisted_components = {}
+
+    def reset(self):
+        self.held_objmaps = {}
+        self.blacklisted_components = {}
+
+    def now(self):
+        return time()
+
+    def append(self, objmap):
+        # for convenience, ignore None objmaps if passed in.  (mapping functions
+        # can return None if there's no data to apply, and it's easier to just
+        # check for that here)
+        if objmap is None:
+            return
+
+        component_id = objmap.id
+
+        if component_id not in self.held_objmaps:
+            self.held_objmaps[component_id] = (self.now(), objmap)
+        else:
+            _, stored_objmap = self.held_objmaps[component_id]
+
+            # Only overwrite _add and remove if the value is True.
+            if getattr(objmap, '_add', False):
+                stored_objmap._add = objmap._add
+            if getattr(objmap, 'remove', False):
```

Comment:
I'd recommend using `_remove_` instead of `remove` throughout as that's the currently-documented approach in PythonCollector. The `remove` attribute is only supported for backwards compatibility.

Reply:
done